libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -682,28 +682,32 @@ string YulUtilFunctions::overflowCheckedIntMulFunction(IntegerType const& _type)
 			function <functionName>(x, y) -> product {
 				x := <cleanupFunction>(x)
 				y := <cleanupFunction>(y)
+				product := mul(x, y)
 				<?signed>
-					// overflow, if x > 0, y > 0 and x > (maxValue / y)
-					if and(and(sgt(x, 0), sgt(y, 0)), gt(x, div(<maxValue>, y))) { <panic>() }
-					// underflow, if x > 0, y < 0 and y < (minValue / x)
-					if and(and(sgt(x, 0), slt(y, 0)), slt(y, sdiv(<minValue>, x))) { <panic>() }
-					// underflow, if x < 0, y > 0 and x < (minValue / y)
-					if and(and(slt(x, 0), sgt(y, 0)), slt(x, sdiv(<minValue>, y))) { <panic>() }
-					// overflow, if x < 0, y < 0 and x < (maxValue / y)
-					if and(and(slt(x, 0), slt(y, 0)), slt(x, sdiv(<maxValue>, y))) { <panic>() }
+					// overflow, if product has positive sign and
+					// y != (p & bitMask) / x
+					if and(
+						iszero(and(product, <signMask>)),
+						and(iszero(iszero(x)), iszero(eq(y, sdiv(and(product, <bitMask>),x))))
+					) { <panic>() }
+					// overflow, if product has negative sign and
+					// y != (p | !bitMask) / x
+					if and(
+						iszero(iszero(and(product, <signMask>))),
+						and(iszero(iszero(x)), iszero(eq(y, sdiv(or(product, not(<bitMask>)),x))))
+					) { <panic>() }
```

Comment:
Seems it works ;) But, I think we can make more gas savings by moving duplicated operations  before that code block:
- `iszero(and(product, <signMask>))`,
- `iszero(iszero(x))` (I guess this one won't help that much).

Reply:
We can move `iszero(iszero(x))` like this (?):
```
if and(
	iszero(iszero(x)),
	or(
		and(
			iszero(and(product, <signMask>)),
			iszero(eq(y, sdiv(and(product, <bitMask>),x)))
		),
		and(
			iszero(iszero(and(product, <signMask>))),
			iszero(eq(y, sdiv(or(product, not(<bitMask>)),x)))
		)
	)
) { <panic>() }
```
 But `iszero(and(product, <signMask>))` is not duplicated. In the second block it is actually negated. I don't think we can remove those.

Reply:
It is possible to exchange `iszero(iszero(and(product, <signMask>))` for `eq(<signMask>, and(product, <signMask>)) `.